clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -223,23 +261,41 @@ public void ensureActiveGroup() {
                 continue;
             }
 
-            RequestFuture<ByteBuffer> future = sendJoinGroupRequest();
-            future.addListener(new RequestFutureListener<ByteBuffer>() {
-                @Override
-                public void onSuccess(ByteBuffer value) {
-                    // handle join completion in the callback so that the callback will be invoked
-                    // even if the consumer is woken up before finishing the rebalance
-                    onJoinComplete(generation, memberId, protocol, value);
-                    needsJoinPrepare = true;
-                    heartbeatTask.reset();
-                }
+            // we store the join future in case we are woken up by the user after beginning the
+            // rebalance in the call to poll below. This ensures that we do not mistakenly attempt
+            // to rejoin before the pending rebalance has completed.
+            if (joinFuture == null) {
+                state = MemberState.REBALANCING;
+                joinFuture = sendJoinGroupRequest();
+                joinFuture.addListener(new RequestFutureListener<ByteBuffer>() {
+                    @Override
+                    public void onSuccess(ByteBuffer value) {
+                        // handle join completion in the callback so that the callback will be invoked
+                        // even if the consumer is woken up before finishing the rebalance
+                        synchronized (AbstractCoordinator.this) {
+                            log.info("Successfully joined group {} with generation {}", groupId, generation.generationId);
+                            joinFuture = null;
+                            state = MemberState.STABLE;
+                            needsJoinPrepare = true;
+                            heartbeatThread.enable();
+                        }
 
-                @Override
-                public void onFailure(RuntimeException e) {
-                    // we handle failures below after the request finishes. if the join completes
-                    // after having been woken up, the exception is ignored and we will rejoin
-                }
-            });
+                        onJoinComplete(generation.generationId, generation.memberId, generation.protocol, value);
+                    }
+
+                    @Override
+                    public void onFailure(RuntimeException e) {
+                        // we handle failures below after the request finishes. if the join completes
+                        // after having been woken up, the exception is ignored and we will rejoin
+                        synchronized (AbstractCoordinator.this) {
+                            joinFuture = null;
+                            state = MemberState.UNJOINED;
+                        }
+                    }
+                });
+            }
+
+            RequestFuture<ByteBuffer> future = joinFuture;
```

Comment:
Don't need to copy the future here. We're synchronized so the couple of lines can be reduced to `client.poll(joinFuture)`.